### ISSUE TITLE
Avoid recomputing unnecessary subtrees

### ISSUE THREAD

**Issue by MaxGhenis** [PolicyEngine/policyengine-core#52]
Based on the cached baseline

cc @michaelsnowden @ibushong

**Comment by anrikus** [comment]
To get more clarity, does `baseline` mean the branch_name is `default`, as in [here](https://github.com/PolicyEngine/policyengine-core/blob/master/policyengine_core/simulations/simulation.py#L84)?

If not, then what is the criterion for determining the `baseline`?

The `branch_name` also shows up with `sim.tracer.print_computation_log()`:

<img width="885" alt="image" src="https://github.com/PolicyEngine/policyengine-core/assets/10928434/416648cc-859e-4494-952b-e86c1d5eaf85">

**Comment by MaxGhenis** [comment]
@nikhilwoodruff

**Comment by anrikus** [comment]
@MaxGhenis @nikhilwoodruff 

[Current State] Here's my understanding so far:
- `Simulation` depends on `variables`.
- `Variables` depend on `parameters` and `reforms`.
- `Reforms` alter or introduce new `variables` and `parameters`.
- Computation graph for simulation is dynamic and computed during runtime (no persistent reference).

[Requirements] To implement caching:
- We need a persistent [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) to represent the `baseline` computation where: 
  - a `node` is a `variable` or a `parameter`
  - `downstream link` represents a direct dependency for the `node`
- Parse `reforms` and compute which `nodes` are altered.
- Invalidate all `altered nodes` and their `upstream nodes`.
- During simulation:
  - Consume the list of `invalid nodes`.
  - Recompute value for all `invalid nodes`.
  - Use cached values for all `unaltered nodes`.

[High level feature work items]:
- Determine a `baseline` scenario.
- Compute and save simulation DAGs for a given configuration on persistent storage.
- Parse `reforms`.
- Determine `invalid nodes` for a given DAG and a given set of `reforms`.
- Consume `invalid nodes` info during runtime and conditionally re-calculate / fetch `node` values. 

It is very likely that my understanding of the problem / implementation / possible solution is wrong (was chasing a different goose till yesterday). Please correct me / point me to the right resources.

**Comment by MaxGhenis** [comment]
This all sounds right, thanks for detailing it. I think this will step will require the most work:
>Parse reforms and compute which nodes are altered.

**Comment by nikhilwoodruff** [comment]
Yes +1. Thanks @anrikus for this. My thinking for an initial prototype that could be useful here is:

* Generate a DAG as you specify, **assuming it's constant for all variable and parameter values**[^1].
* Cache the DAG for a given variable and parameter input value set (the baseline).
* For a given different input set, identify the nodes in the tree that need recalculating


[^1]: This assumption is not true. But if we find _significant_ speedup opportunities using it, we can make it true by adjusting the country package codebases.
